sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitmapExpressions.scala

```diff
@@ -248,9 +258,17 @@ case class BitmapOrAgg(child: Expression,
 
   override def checkInputDataTypes(): TypeCheckResult = {
     if (child.dataType != BinaryType) {
-      TypeCheckResult.TypeCheckFailure("Bitmap must be a BinaryType")
+      DataTypeMismatch(
```

Comment:
Should we mark `TypeCheckFailure` as `deprecated` in the Spark 3.5 code?

Reply:
+1. I checked the code. This PR will remove the last two apply for `TypeCheckResult.TypeCheckFailure`.
cc @MaxGekk @cloud-fan

Reply:
I know many people implement custom expressions, and removing `TypeCheckFailure` may break them. I think it's OK to leave it and we can just remove all its usages in Spark.

Reply:
> I know many people implement custom expressions, and removing `TypeCheckFailure` may break them. I think it's OK to leave it and we can just remove all its usages in Spark.

Can we make it as `deprecated ` first?

Reply:
In general, we don't encourage deprecation if we are not going to remove it. cc @zhengruifeng

Reply:
+1, let's not deprecate it

Reply:
So, what measures can we take to prevent it from being used again in the Spark code? It seems to be discouraged.’

Reply:
```
Why are the changes needed?
It appears TypeCheckResult.TypeCheckFailure has been deprecated: No expressions except for the recently added BitmapCount and BitmapOrAgg are using it.
```

@bersprockets On the other hand, even though the pr has been merged, could you modify the PR description? From the suggestions of @cloud-fan  and @zhengruifeng , there is currently no intention to make TypeCheckFailure deprecated.

Reply:
I think we should use deprecation only when we have to remove it in the future.
An example is the Pandas 2.0 Upgrade, Pandas 2.0 has introduced a lot of breaking changes, some APIs/parameters were removed. So in Pandas API on Spark, we will have to make similar changes for Pandas parity, e.g. 

https://github.com/apache/spark/blob/a367fde24de0abab93eac97350fb4ae0b687286c/python/pyspark/pandas/frame.py#L2060-L2072

https://github.com/apache/spark/blob/20a8fc87d67c842ac3386dc6ae0c53a9533900c2/python/pyspark/pandas/series.py#L1684-L1687

Reply:
OK, let’s continue to keep it.